### PR TITLE
Reuse PKCS11 objects for keys

### PR DESCRIPTION
**Pull request by mastersingh24**
#### Type of change

- Improvement (improvement to code, performance, etc)

#### Description

Cherry pick #1669 to release-2.2

### CODE REVIEW

File: bccsp/pkcs11/impl.go [hunk 1]
```diff
@@ -151,14 +175,28 @@ func (csp *impl) KeyImport(raw interface{}, opts bccsp.KeyImportOpts) (k bccsp.K
...
-	pubKey, isPriv, err := csp.getECKey(ski)
-	if err == nil {
-		if isPriv {
-			return &ecdsaPrivateKey{ski, ecdsaPublicKey{ski, pubKey}}, nil
-		}
-		return &ecdsaPublicKey{ski, pubKey}, nil
+	privHandle, pubHandle, pubKey, err := csp.getECKey(ski)
+	if err != nil {
+		logger.Warnf("Key not found using PKCS11: %v", err)
+		return csp.BCCSP.GetKey(ski)
```
Comment (sykesm) [inline bccsp/pkcs11/impl.go:181]: I expect this is going to happen quite frequently.

**Review by lindluni — DISMISSED** [review]

**Review by sykesm — COMMENTED** [review]
With this change, handles are held in the public and private key implementation.

Does this need additional paths to deal with stale handles? What happens if all cached sessions are purged?

**Review by sykesm — CHANGES_REQUESTED** [review]
I'm going to put this in a holding pattern until @mastersingh24 can respond.

File: bccsp/pkcs11/impl.go (hunk 1 above)
Comment (lindluni) [reply bccsp/pkcs11/impl.go:181]: I introduced this after Mihir and I were trying to debug an issue but couldn't trace the failure as none of there errors were logged and Gari picked it up as part of his cherry-pick. Should probably be raised in severity into `debug`. I'll make the change in master and cherry-pick it.

**Comment by sykesm** [comment]
Looks like there may be something to my concerns. I made the following change to emulate what would happen if all sessions in the cache were purged due to some error:

```diff
diff --git a/bccsp/pkcs11/impl.go b/bccsp/pkcs11/impl.go
index 49b6bf26d..13a818415 100644
--- a/bccsp/pkcs11/impl.go
+++ b/bccsp/pkcs11/impl.go
@@ -20,7 +20,7 @@ import (

 var (
        logger           = flogging.MustGetLogger("bccsp_p11")
-       sessionCacheSize = 10
+       sessionCacheSize = 1
 )

 // New WithParams returns a new instance of the software-based BCCSP
diff --git a/bccsp/pkcs11/pkcs11.go b/bccsp/pkcs11/pkcs11.go
index 9fba499fd..ac11912b5 100644
--- a/bccsp/pkcs11/pkcs11.go
+++ b/bccsp/pkcs11/pkcs11.go
@@ -110,6 +110,7 @@ func createSession(ctx *pkcs11.Ctx, slot uint, pin string) (pkcs11.SessionHandle
 }

 func (csp *impl) returnSession(session pkcs11.SessionHandle) {
+       csp.ctx.CloseSession(session)
        select {
        case csp.sessions <- session:
                // returned session back to session cache
```

Before caching the handles, `integration/pkcs11` would run successfully; after caching the handles it fails and panics:

```
[e][Org2.peer0] 2020-08-03 09:39:27.456 EDT [nodeCmd] createSelfSignedData -> PANI 006 Failed creating self signed data because message signing failed: Sign-initialize  failed [pkcs11: 0x82: CKR_OBJECT_HANDLE_INVALID]
[e][Org1.peer0] panic: Failed creating self signed data because message signing failed: Sign-initialize  failed [pkcs11: 0x82: CKR_OBJECT_HANDLE_INVALID]
[e][Org1.peer0]
[e][Org1.peer0] goroutine 1 [running]:
[e][Org1.peer0] go.uber.org/zap/zapcore.(*CheckedEntry).Write(0xc000225e40, 0x0, 0x0, 0x0)
[e][Org1.peer0] 	/Users/sykesm/workspace/fabric/src/github.com/hyperledger/fabric/vendor/go.uber.org/zap/zapcore/entry.go:230 +0x545
[e][Org1.peer0] go.uber.org/zap.(*SugaredLogger).log(0xc0001380f8, 0xc0004b1904, 0x500c01a, 0x43, 0xc00019ec10, 0x1, 0x1, 0x0, 0x0, 0x0)
[e][Org1.peer0] 	/Users/sykesm/workspace/fabric/src/github.com/hyperledger/fabric/vendor/go.uber.org/zap/sugar.go:234 +0x100
[e][Org1.peer0] go.uber.org/zap.(*SugaredLogger).Panicf(...)
[e][Org1.peer0] 	/Users/sykesm/workspace/fabric/src/github.com/hyperledger/fabric/vendor/go.uber.org/zap/sugar.go:159
[e][Org1.peer0] github.com/hyperledger/fabric/common/flogging.(*FabricLogger).Panicf(...)
[e][Org1.peer0] 	/Users/sykesm/workspace/fabric/src/github.com/hyperledger/fabric/common/flogging/zap.go:74
[e][Org1.peer0] github.com/hyperledger/fabric/internal/peer/node.createSelfSignedData(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
[e][Org1.peer0] 	/Users/sykesm/workspace/fabric/src/github.com/hyperledger/fabric/internal/peer/node/start.go:919 +0x187
[e][Org1.peer0] github.com/hyperledger/fabric/internal/peer/node.serve(0x5bdc6d8, 0x0, 0x0, 0x0, 0x0)
[e][Org1.peer0] 	/Users/sykesm/workspace/fabric/src/github.com/hyperledger/fabric/internal/peer/node/start.go:231 +0x32b
[e][Org1.peer0] github.com/hyperledger/fabric/internal/peer/node.glob..func6(0x5b88640, 0x5bdc6d8, 0x0, 0x0, 0x0, 0x0)
[e][Org1.peer0] 	/Users/sykesm/workspace/fabric/src/github.com/hyperledger/fabric/internal/peer/node/start.go:127 +0x9c
[e][Org1.peer0] github.com/spf13/cobra.(*Command).execute(0x5b88640, 0x5bdc6d8, 0x0, 0x0, 0x5b88640, 0x5bdc6d8)
[e][Org1.peer0] 	/Users/sykesm/workspace/fabric/src/github.com/hyperledger/fabric/vendor/github.com/spf13/cobra/command.go:762 +0x453
[e][Org1.peer0] github.com/spf13/cobra.(*Command).ExecuteC(0x5b88d60, 0xc000375f50, 0x1, 0x1)
[e][Org1.peer0] 	/Users/sykesm/workspace/fabric/src/github.com/hyperledger/fabric/vendor/github.com/spf13/cobra/command.go:852 +0x2ea
[e][Org1.peer0] github.com/spf13/cobra.(*Command).Execute(...)
[e][Org1.peer0] 	/Users/sykesm/workspace/fabric/src/github.com/hyperledger/fabric/vendor/github.com/spf13/cobra/command.go:800
[e][Org1.peer0] main.main()
[e][Org1.peer0] 	/Users/sykesm/workspace/fabric/src/github.com/hyperledger/fabric/cmd/peer/main.go:54 +0x46d
[e][Org2.peer0] panic: Failed creating self signed data because message signing failed: Sign-initialize  failed [pkcs11: 0x82: CKR_OBJECT_HANDLE_INVALID]
```

**Comment by mastersingh24** [comment]
@sykesm  - thanks ... will have a look ... looks like a misread the actual pkcs11 spec ... looks like all sessions can access the same object handle but if the initial session where login occurred goes away, the object handle is no longer valid.
Will dig into the code and see if there are any better options.

**Pull request closed without merging**